### PR TITLE
Fix Executor handler validation for postponed annotations (future annotations)

### PR DESCRIPTION
**Pull request by moonbox3**
## Problem
When `from __future__ import annotations` is enabled, Executor handler parameter annotations are strings. `_validate_handler_signature` passed these strings into `validate_workflow_context_annotation`, causing `WorkflowContext[T, U]` validation to fail with `ValueError`.

## Fix
- Resolve handler annotations inside `_validate_handler_signature` using `typing.get_type_hints` (best-effort; falls back to raw annotations on resolution errors).
- Validate `WorkflowContext[...]` using the resolved type hint instead of the raw string.
- Store the resolved `ctx_annotation` in the handler spec (including explicit-type mode).

## Tests
Added `test_executor_future.py` to ensure that handlers defined under postponed annotations validate successfully and that `output_types` / `workflow_output_types` are inferred correctly for:
- `WorkflowContext[TypeA, TypeB]`
- `WorkflowContext[TypeA | TypeB, Never]`

### CODE REVIEW

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 1]
```diff
@@ -717,26 +715,46 @@ def _validate_handler_signature(
     if len(params) != expected_counts:
         raise ValueError(f"Handler {func.__name__} must have {param_description}. Got {len(params)} parameters.")
 
+    # Best-effort resolve postponed annotations (from __future__ import annotations)
+    # so downstream validators see real typing objects rather than strings.
+    try:
+        import typing
+
+        type_hints = typing.get_type_hints(func, include_extras=True)
+    except TypeError:
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:724]: `get_type_hints(func, include_extras=True)` is called without explicit `globalns/localns`; for nested/local handler definitions or forward refs this can raise and you then fall back to `{}`, reintroducing raw/string annotations. Pass appropriate namespaces (e.g., `func.__globals__` plus a suitable `localns`) and/or fail fast with a clear error when ctx annotation is required in introspection mode.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 2]
```diff
@@ -717,26 +715,46 @@ def _validate_handler_signature(
...
+        # include_extras not supported
+        try:
+            import typing
+
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:728]: All `get_type_hints` errors are currently swallowed and you fall back to raw annotations (strings under postponed annotations). Add coverage that forces `get_type_hints` to fail (undefined forward ref) and ensure behavior is controlled (clear ValueError or explicit-type mode continues) rather than silently regressing.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 3]
```diff
@@ -717,26 +715,46 @@ def _validate_handler_signature(
...
+        message_type = None
+
     # Validate ctx parameter is WorkflowContext and extract type args
     ctx_param = params[2]
-    if skip_message_annotation and ctx_param.annotation == inspect.Parameter.empty:
+    ctx_annotation = type_hints.get(ctx_param.name, ctx_param.annotation)
+
+    if skip_message_annotation and ctx_annotation == inspect.Parameter.empty:
         # When explicit types are provided via @handler(input=..., output=...),
         # the ctx parameter doesn't need a type annotation - types come from the decorator.
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:750]: When `skip_message_annotation=True` and ctx is missing, `ctx_annotation` can remain `inspect.Parameter.empty` (or a string via fallback) and then be stored/returned. Normalize to `None` (or another stable typing value) before storing in handler specs so downstream code doesn’t need to special-case inspect sentinels.

File: python/packages/core/tests/workflow/test_executor_future.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) Microsoft. All rights reserved.
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future.py:1]: Add tests for explicit-type decorator mode under postponed annotations: `@handler(input=..., output=...)` with (a) ctx unannotated and (b) ctx annotated (postponed) to verify `resolved_ctx_annotation` handling and that output/workflow-output types come from the decorator as intended.

File: python/packages/core/tests/workflow/test_executor_future.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        assert set(ex.workflow_output_types) == {TypeB}
+
+    def test_handler_future_annotations_workflow_context_union_args(self) -> None:
+        class FutureUnionArgsExecutor(Executor):
+            @handler
+            async def handle(self, message: str, ctx: WorkflowContext[TypeA | TypeB, Never]) -> None:
+                pass
+
+        ex = FutureUnionArgsExecutor(id="future_union")
+        assert set(ex.output_types) == {TypeA, TypeB}
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future.py:41]: The assertion `workflow_output_types == []` for `WorkflowContext[..., Never]` is brittle across Python/typing variants (some may produce `[Never]`). Update the validator/tests to define and assert the intended `Never` semantics consistently (normalize to no outputs, or accept the normalized representation).

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Issue:** #1 | **Reviewers:** 3 | **Confidence:** 74%

### Summary
The PR’s approach of resolving postponed annotations with `typing.get_type_hints` appears to address the `from __future__ import annotations` regression, and the added tests cover the basic introspection scenario. However, there are reliability and contract gaps: `get_type_hints` is used without namespaces and failures are silently masked, and tests miss explicit-type decorator mode and key failure/edge cases. Changes are needed to ensure consistent resolution/normalization and robust, non-brittle test expectations across Python/typing implementations.

### Blocking Issues
- [ ] Add regression tests for explicit-type decorator mode (`@handler(input=..., output=...)`) under postponed annotations, including cases where ctx is unannotated and where ctx is annotated (postponed) to ensure `resolved_ctx_annotation`/ctx sourcing behaves correctly.
- [ ] Make `get_type_hints` resolution reliable: it is currently called without explicit `globalns/localns` and errors are swallowed, which can reintroduce string/empty annotations and inconsistent behavior (especially for nested/local definitions or forward refs). Either pass appropriate namespaces and/or fail fast with a clear error when ctx annotation is required.
- [ ] Normalize ctx annotation output: when `skip_message_annotation=True` and ctx is missing/unresolvable, the code can propagate `inspect.Parameter.empty` (or strings via fallback) into the handler spec; convert to a stable representation (e.g., `None` or a resolved typing object) so downstream consumers don’t have to special-case inspect sentinels.
- [ ] Fix/clarify the test contract around `WorkflowContext[..., Never]`: asserting `workflow_output_types == []` is brittle across typing/Python variants; update validator/tests to consistently define and assert the intended semantics for `Never` (e.g., normalize to no outputs, or accept the normalized representation).

### Suggestions
- Add a test that forces `get_type_hints` to fail (e.g., undefined forward ref) and assert behavior is controlled (clear ValueError when ctx is required, or successful explicit-type mode when ctx is optional).
- Extend tests to assert `message_type` is also resolved correctly under postponed annotations, not just ctx/output extraction.
- Narrow exception handling around `get_type_hints` to avoid silently masking resolution problems; surface actionable error messages when introspection mode cannot resolve required annotations.

---
*Automated review by triage-bot*

**Comment by moonbox3** [comment]
## Review Feedback Update

All four previously blocking issues appear addressed: explicit-type decorator mode now has regression coverage under postponed annotations, type-hint resolution uses explicit namespaces and fails fast when required, ctx annotation is normalized to None in explicit mode, and the Never workflow-output contract in tests is no longer brittle.

Changes pushed in commit `f29056fa7c28eb8972ed88a017c72c4bc6bf139b`.

---
*Automated review by triage-bot*
